src/v/cluster/rm_stm.cc

```diff
@@ -1069,35 +1071,183 @@ ss::future<result<raft::replicate_result>> rm_stm::replicate_seq(
           "first_seq={} of the batch should be less or equal to last_seq={}",
           bid.first_seq,
           bid.last_seq);
-        outcome->set_value();
         co_return errc::generic_tx_error;
     }
 
+    // getting session by client's pid
+    ss::lw_shared_ptr<inflight_requests> session;
+    auto session_it = _inflight_requests.find(bid.pid);
```

Comment:
nit: can be replaced with a code doing a single lookup:
```c++
  auto [session_it, _] = _inflight_requests.try_emplace(
      bid.pid, ss::make_lw_shared<inflight_requests>());

    auto session = session_it->second;
```

Reply:
```c++
    auto [session_it, _] = _inflight_requests.try_emplace(bid.pid);
    auto session = session_it->second;
```

Reply:
a pointer isn't set by default

Reply:
Sometimes `.lazy_emplace` is the correct tool. Last time I checked, that code hadn't changed much, I doubt it's going anywhere. I can't see a way to implement it as a free function with the same efficiency.

Reply:
Thanks, Ben, I've changed to `lazy_emplace`. @mmaslankaprv we found a compromise!